pkg/logql/shardmapper.go

```diff
@@ -5,36 +5,147 @@ import (
 )
 
-func NewShardMapper(shards int) (ShardMapper, error) {
+// keys used in metrics
+const (
+	StreamsKey = "streams"
+	MetricsKey = "metrics"
+	SuccessKey = "success"
+	FailureKey = "failure"
+	NoopKey    = "noop"
+)
+
+// ShardingMetrics is the metrics wrapper used in shard mapping
+type ShardingMetrics struct {
+	shards      *prometheus.CounterVec // sharded queries total, partitioned by (streams/metric)
+	parsed      *prometheus.CounterVec // parsed ASTs total, partitioned by (success/failure/noop)
+	shardFactor prometheus.Histogram   // per request shard factor
+}
+
+func NewShardingMetrics(registerer prometheus.Registerer) *ShardingMetrics {
+
+	return &ShardingMetrics{
+		shards: promauto.With(registerer).NewCounterVec(prometheus.CounterOpts{
+			Namespace: "loki",
+			Name:      "query_frontend_shards_total",
+		}, []string{"type"}),
+		parsed: promauto.With(registerer).NewCounterVec(prometheus.CounterOpts{
+			Namespace: "loki",
+			Name:      "query_frontend_sharding_parsed_queries_total",
+		}, []string{"type"}),
+		shardFactor: promauto.With(registerer).NewHistogram(prometheus.HistogramOpts{
+			Namespace: "loki",
+			Name:      "query_frontend_shard_factor",
+			Help:      "Number of shards per request",
+			Buckets:   prometheus.LinearBuckets(0, 16, 4), // 16 is the default shard factor for later schemas
+		}),
+	}
+}
+
+// ShardRecorder constructs a recorder using the underlying metrics.
+func (m *ShardingMetrics) ShardRecorder() *ShardRecorder {
+	return &ShardRecorder{
+		ShardingMetrics: m,
+	}
+}
+
+// ShardRecorder wraps a vector & histogram, providing an easy way to increment sharding counts.
+// and unify them into histogram entries.
+// NOT SAFE FOR CONCURRENT USE! We avoid introducing mutex locking here
+// because AST mapping is single threaded.
+type ShardRecorder struct {
+	done  bool
+	total int
+	*ShardingMetrics
+}
+
+// Add increments both the shard count and tracks it for the eventual histogram entry.
+func (r *ShardRecorder) Add(x int, key string) {
+	r.total += x
+	r.shards.WithLabelValues(key).Add(float64(x))
+}
+
+// Finish idemptotently records a histogram entry with the total shard factor.
+func (r *ShardRecorder) Finish() {
+	if !r.done {
+		r.done = true
+		r.shardFactor.Observe(float64(r.total))
+	}
+}
+
+func badASTMapping(expected string, got Expr) error {
+	return fmt.Errorf("Bad AST mapping: expected one type (%s), but got (%T)", expected, got)
+}
+
+func NewShardMapper(shards int, metrics *ShardingMetrics, log log.Logger) (ShardMapper, error) {
 	if shards < 2 {
 		return ShardMapper{}, fmt.Errorf("Cannot create ShardMapper with <2 shards. Received %d", shards)
 	}
-	return ShardMapper{shards}, nil
+	return ShardMapper{
+		shards:  shards,
+		metrics: metrics,
+		logger:  log,
+	}, nil
 }
 
 type ShardMapper struct {
-	shards int
+	shards  int
+	metrics *ShardingMetrics
+	logger  log.Logger
+}
+
+func (m ShardMapper) Parse(query string) (Expr, error) {
+	parsed, err := ParseExpr(query)
```

Comment:
I really appreciate the instrumentation you've done for the mapper this is going to be really useful.